src/ssh/cockpitsshrelay.c

```diff
@@ -1320,61 +1320,61 @@ send_auth_reply (CockpitSshData *data,
   return ret;
 }
 
-static void
+static gboolean
 parse_host (const gchar *host,
             gchar **hostname,
             gchar **username,
             guint *port)
 {
-  gchar *user_arg = NULL;
-  gchar *host_arg = NULL;
-  gchar *tmp = NULL;
-  gchar *end = NULL;
-
-  guint64 tmp_num;
-
-  gsize host_offset = 0;
-  gsize host_length = strlen (host);
-
-  tmp = strrchr (host, '@');
-  if (tmp)
-    {
-      if (tmp[0] != host[0])
+  GError *error = NULL;
+  g_autoptr (GRegex) regex = g_regex_new ("^"
+         "(?:(.+)@)?"         /* optional username */
+         "(?|"                /* one of... */
+           "\\[([^]@]+)\\]"     /* hostname in square brackets, no @ */
+           "(?::([1-9][0-9]*))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@:]+)"           /* hostname with no : or @ */
+           "(?::([1-9][0-9]*))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@]+)"            /* hostname with no @ but : (IPv6 address), and no port */
+         ")"                  /* . */
+       "$",
+       0, 0, &error);
+  g_assert_no_error (error);
+
+  g_autoptr(GMatchInfo) info = NULL;
+
+  if (g_regex_match (regex, host, 0, &info))
+    {
+      g_autofree gchar *port_str = g_match_info_fetch (info, 3);
+      /* regexp makes sure that it's a positive number, so don't need much error checking */
+      guint value = atoi (port_str ?: "");
+      if (value >= 0 && value < 65536)
```

Comment:
`value` is unsigned, so it's always `>= 0`.  I did the `guint` for that reason.  You could also turn it into a `gint` (to better match the return type of `atoi()`) if you wanted, and then this extra check would make sense...

Reply:
Argh, indeed -- tests didn't get very far yet, I just fixed it.